util/vec-map/src/lib.rs

```diff
@@ -0,0 +1,300 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Provides a map (key-value store) interface backed by `heapless::Vec`
+
+#![deny(missing_docs)]
+
+use displaydoc::Display;
+use heapless::Vec;
+
+/// An error which can occur when using VecMap
+#[derive(Clone, Debug, Display, Eq, PartialEq)]
+pub enum Error {
+    /// VecMap capacity exceeded
+    CapacityExceeded,
+}
+
+/// This is a mini version of VecMap that is no_std compatible and uses
+/// heapless::Vec instead alloc::Vec.
+///
+/// It may be better to patch upstream for no_std compatibility and use that,
+/// but that crate has other issues -- it relies on an
+/// experimental "contracts" crate that causes a dependency on rand crate.
+/// Porting to Heapless would be a breaking change to the API.
+///
+/// TBD what the best path is: https://github.com/p-avital/vec-map-rs/blob/master/src/lib.rs
+#[derive(Clone, Debug)]
+pub struct VecMap<K, V, const N: usize> {
+    keys: Vec<K, N>,
+    values: Vec<V, N>,
```

Comment:
Is the reason to go with separate keys/values as opposed to `Vec<(K, V), N>` has to do with the borrow checker? Or is there any other reason for this?

Reply:
* It can potentially be faster to search if the keys are small and values are big, since you dont have to page in all the values into memory when you scan
* It can be smaller on the stack, if for example the key is a u64 and has alignnent of 8, and the value is a byte buffer not divisible by 8 or something else not aligned to 8 bytes

Reply:
See also what this guy wrote: https://crates.io/crates/vector-map